Adding Cosine and Jaccard (#49)
lennessyy/graph-ml: modules/similarity-algorithms/pages/cosine-similarity.adoc

= Cosine Similarity
:stem: latex
:experimental:

> Algorithm link: link:https://github.com/tigergraph/gsql-graph-algorithms/tree/277349ce0414ba797edcad1488b6caf1904d5beb/algorithms/Similarity/approximate_nearest_neighbors[Approximate Nearest Neighbors]

Reviewer: This is a UDF function, not an algorithm.  Note the location of the UDF file, specified in the updated DOC ticket.

Author: Changed this.

Cosine similarity measures the cosine of the angle between two non-zero vectors of an inner product space.

[stem]
++++
\[
\text{Cosine Similarity} = \frac{\sum_{i=1}^{n}A_i \times B_i}{\sqrt{\sum_{i=1}^{n}A_i^2} \times \sqrt{\sum_{i=1}^{n}B_i^2}}
\]
++++

> This is commonly used to determine the similarity between two vectors, often used in text analysis and other vector space models.

Reviewer: ```suggestion
If each entity is being modeled as a set of numerically scored characteristics, than cosine similarity is a commonly used method of measuring the similarity between two entities. 
```

Author: Thank you. Added with latest commit.

> The algorithm can take vectors denoted by `ListAccum` or standard `vector`.

Reviewer: ```suggestion
The function operates on vectors formatted as `ListAccum` accumulators.
```

Author: Thank you. Added with latest commit.

== Notes

This algorithm is implemented as a user-defined function.
You need to follow the steps in xref:3.10.1@gsql-ref:querying:func/query-user-defined-functions.adoc[Add a User-Defined Function] to add the function to GSQL.

> After adding the function, you can call it in any GSQL query in the same way as a built-in GSQL function.

Reviewer: ```suggestion
The cosine similarity function is implemented as one option of a multifunction user-defined function (UDF) that is preinstalled in GSQL.
```

Author: Thank you changed with latest commit.

== Specifications

....

> tg_cosine_similarity_accum(A, B)
> tg_cosine_similarity(A, B)

Reviewer: ```suggestion
tg_similarity_accum(A, B, "COSINE")
```

Author: Changed this for both Cosine and Jaccard.

....

=== Parameters

[options="header",]
|===
|Name |Description |Data type
|`+A+`
|An n-dimensional vector denoted by a `+ListAccum+` or `+vector+` of length `+n+`
|`+ListAccum<INT/UINT/FLOAT/DOUBLE>+` or `+vector<INT/UINT/FLOAT/DOUBLE>+`

|`+B+`
|An n-dimensional vector denoted by a `+ListAccum+` or `+vector+` of length `+n+`
|`+ListAccum<INT/UINT/FLOAT/DOUBLE>+` or `+vector<INT/UINT/FLOAT/DOUBLE>+`

> |===

Reviewer: Add the function selector parameter.

Author: Thank you added to all references of vector functions.

=== Output

The cosine similarity between the two vectors, which ranges from -1 (exactly opposite) to 1 (exactly the same), with 0 indicating orthogonality.

=== Time complexity

The algorithm has a complexity of stem:[O(n)], where stem:[n] is the number of dimensions of the vectors.

== Example
[tabs]
====
Accum Query::
+
--
[,gsql]
----
CREATE QUERY cosine_similarity_example() FOR GRAPH social {
  ListAccum<INT> @@a = [1, 0, 3];
  ListAccum<INT> @@b = [0, 2, 6];

>   double similarity = tg_cosine_similarity_accum(@@a, @@b);

Reviewer: ```suggestion
  double similarity = tg_similarity_accum(@@a, @@b, "COSINE");
```

Author: This has been changed with all referenced vector functions.

  PRINT similarity;
}
----
--
Vector Query::
+
--
[,gsql]
----
CREATE QUERY cosine_similarity_std_example() FOR GRAPH social {
  std::vector<int> a = {1, 0, 3};
  std::vector<int> b = {0, 2, 6};
  double similarity = tg_cosine_similarity(a, b);
  PRINT similarity;
}
----

> --

Reviewer: ```suggestion
```

Author: Thank you. I removed these.
